### DISCUSSION TITLE
Convert sfnetwork to linnet while preserving edge geometries

### DISCUSSION THREAD

**Discussion by mkvasnicka** [luukvdmeer/sfnetworks#207] [Q&A]
Hi!

I work with spatstat's linnet and derived classes, but I want to preprocess my data with sfnetworks. (I'm working with a road network.) Therefore, I need a way to convert a sfnetwork to linnet. I've stumbled across three ways to do it:

First, one can use as.linnet.sfnetwork(). This is fast and preserves the topology but doesn't work for my purposes as it replaces each edge with a straight line, dropping all spatial information about the paths.

Second, one can convert the edges to sf, sf to sp, and sp to linnet. Here, I found two solutions. One is

```{r}
sfnet |>
        # convert sfnetwork edges to sf
        activate(edges) |>
        st_as_sf() |>
        st_geometry() |>
        # transform to sp
        as_Spatial() |>
        # transform to psp
        as.psp() |>
        # transform to linnet
        as.linnet.psp(sparse = TRUE)
```

The problem with this solution is that it creates "lifts" in mixmasters, i.e., it creates new nodes and connects paths that weren't previously connected but geometrically crossed. This is happening, e.g., in off-grade intersections. I have no idea why it doesn't work correctly.

The other is very similar---only sp to linnet conversion differs:

```{r}
    sfnet |>
        # convert sfnetwork edges to sf
        activate(edges) |>
        st_as_sf() |>
        st_geometry() |>
        # transform to crs
        st_transform(crs = crs) |>
        # transform to sp
        as_Spatial() |>
        # transform to linnet
        maptools::as.linnet.SpatialLines()
```
This one seems to work ok.

I suggest mentioning this stuff in a vignette for other users.

Please let me know if the third solution is problematic, too, or if there is still a better way.

And many thanks to all developers of sfnetwork for their excellent work.

Best wishes,
Michal

**Comment by agila5** [comment]
Hi @mkvasnicka and thanks for your message. Just to be clear, the problem is that something like

``` r
# packages
library(spatstat)
library(sf)
library(sfnetworks)

# toy network
toy_segment = st_sfc(st_linestring(
  rbind(c(0, 0), c(1, 0), c(1, 1), c(0, 1))
))
toy_sfnet = as_sfnetwork(toy_segment)
plot(toy_sfnet, graticule = TRUE, axes = TRUE, cex = 2)
```

![](https://i.imgur.com/LUjUbAx.png)

is converted to a vertical straight segment from (0, 0) to (0, 1) when converted into spatstat format

``` r
par(mar = rep(0, 4))
plot(as.linnet(toy_sfnet), main = "", vertices = TRUE)
```

![](https://i.imgur.com/hH6wTLg.png)

since you lose the intermediate segments, right?

<sup>Created on 2022-05-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by mkvasnicka** [reply]
Hi. Yes, this is precisely the problem. (In fact, this behavior may be desirable in some situations, e.g., when the network topology is the only thing that matters. However, it is disastrous behavior in my case---I have to make a lot of calculations on a simplified network.)

**Reply by agila5** [reply]
Ok, thanks! I think the ideal solution would be to develop a new spatial morpher to perform exactly that operation. It shouldn't be extremely difficult (since a similar operation is also performed by `to_spatial_subdivision`). I will try to add a PR here as soon as possible.

**Reply by agila5** [reply]
Just for reference, I think that this operation in POSTGIs is called DumpSegments (https://postgis.net/docs/ST_DumpSegments.html) so we could create a morpher with a similar name

**Reply by agila5** [reply]
A possible solution is implemented here: https://github.com/luukvdmeer/sfnetworks/pull/210

**Comment by mkvasnicka** [comment]
Great! Many thanks.

Best wishes,
Michal


Dne pá 27. 5. 2022 17:29 uživatel Andrea Gilardi ***@***.***>
napsal:

**Comment by luukvdmeer** [comment]
In v1.0, you can now run `to_spatial_subdivision()` with the argument `all = TRUE`. This will subdivide edges at each interior point, meaning that every segment becomes its own edge.

```r
as_sfnetwork(roxel) |>
  convert(to_spatial_subdivision, all = TRUE) |>
  as.linnet()
```
